### ISSUE TITLE
Windows:  Use of popup menu created through raw Win32 API causes UI using Slint widgets to go "mute"

### ISSUE THREAD

**Issue by npwoods** [slint-ui/slint#5863]
I've found a problem where creating a popup menu with old-fashioned raw Win32 APIs.  Specifically after `TrackPopupMenu` is called, the UI seems to go "mute" - while mute the UI seems to be functional, but never actually redraws.

I've attached a demonstration of this problem.  In the attached program, there are two buttons:
* One updates a text field, incrementing it by one all of the time
* The other button presents a Win32 popup menu (created with raw Win32 APIs) to provide the option of incrementing by one, two or three

When the popup menu is uses, the text field fails to update.  That said, we can tell that the UI is still "functional" because if you resize the window, the UI stops being mute and the text immediately updates.  Furthermore if you hit the left bump button while the UI is mute, it seems to track them.

Furthermore, the normal UX button highlighting fails while the UI is mute.

Of course, everything about this report is only relevant on Windows.
[win32_popup_problem.zip](https://github.com/user-attachments/files/16641092/win32_popup_problem.zip)

**Comment by ogoffart** [comment]
This may be an issue in winit.

**Comment by npwoods** [comment]
I'm curious if we have any idea what this might be.

I did a lot of stepping through both Slint and Winit to see if I could figure out why a vanilla Win32 context menu would introduce problems, but it exceeded my time box.

**Comment by npwoods** [comment]
I've came up with a gross hack that seems to unfreeze Slint after these popup menus.  Specifically I invoke the following Win32 calls.  The act of changing the windows sizes seems to be important; it is not enough to call `SetWindowPos()` with the existing size.

````
// get the HWND's rectangle
let mut rect = zeroed();
GetWindowRect(hwnd, &mut rect);

// make the window a single pixel wider - the act of changing the size
// seems to "tickle" Slint into unfreezing
SetWindowPos(
	hwnd,
	0 as HWND,
	rect.left,
	rect.top,
	rect.right - rect.left + 1,
	rect.bottom - rect.top,
	0,
);

// and restore the old size
SetWindowPos(
	hwnd,
	0 as HWND,
	rect.left,
	rect.top,
	rect.right - rect.left,
	rect.bottom - rect.top,
	0,
);
````

**Comment by npwoods** [comment]
An interesting finding; if I specify `uFlags` as `SWP_NOMOVE | SWP_NOACTIVATE | SWP_NOOWNERZORDER | SWP_NOSENDCHANGING`, the fix continues to work.  That said, if I specify `SWP_NOSIZE` or `SWP_NOREDRAW`, the fix stops working.

It is not an overall surprise that `SWP_NOSIZE` causes the hack to fail, but the way `SWP_NOREDRAW` spoils the hack is quite interesting.

**Comment by npwoods** [comment]
Honestly I don't know.

I'm a veteran Windows developer that has been in the industry over 25 years, and I tried to debug this issue through Slint and Winit without success.  Nothing seems to be blocked in any particular Windows API (we still get Win32 events) but for some reason nothing seems to emerge from Slint's property engine.  It is in this property engine where I've consistently found myself lost.

My best guess is that somehow, something in Slint and/or Winit gets into a state where events are swallowed or held back (perhaps some sort of reentrancy guard).   And for some reason the specific hack that I crafted must reset out this state.

Are you able to reproduce the problem with my demo?

**Comment by npwoods** [comment]
I took another quick look at this.  Here is a stack trace captured when Slint UI elements are redrawn:

````
>	win32_popup_problem.exe!i_slint_backend_winit::winitwindowadapter::impl$2::request_redraw(i_slint_backend_winit::winitwindowadapter::WinitWindowAdapter * self) Line 724	Rust
 	win32_popup_problem.exe!i_slint_core::window::impl$2::notify(core::pin::Pin<ref$<i_slint_core::window::WindowRedrawTracker>> self) Line 362	Rust
 	win32_popup_problem.exe!i_slint_core::properties::impl$15::new_with_dirty_handler::mark_dirty<i_slint_core::window::WindowRedrawTracker>(i_slint_core::properties::BindingHolder<tuple$<>> * _self, bool was_dirty) Line 1504	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty::closure$0(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 *, i_slint_core::properties::BindingHolder<tuple$<>> * * binding) Line 721	Rust
 	win32_popup_problem.exe!i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>>::for_each<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>,i_slint_core::properties::mark_dependencies_dirty::closure_env$0>(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 self) Line 195	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty(i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>> * dependencies) Line 735	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty::closure$0(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 *, i_slint_core::properties::BindingHolder<tuple$<>> * * binding) Line 734	Rust
 	win32_popup_problem.exe!i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>>::for_each<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>,i_slint_core::properties::mark_dependencies_dirty::closure_env$0>(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 self) Line 195	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty(i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>> * dependencies) Line 735	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty::closure$0(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 *, i_slint_core::properties::BindingHolder<tuple$<>> * * binding) Line 734	Rust
 	win32_popup_problem.exe!i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>>::for_each<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>,i_slint_core::properties::mark_dependencies_dirty::closure_env$0>(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 self) Line 195	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty(i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>> * dependencies) Line 735	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty::closure$0(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 *, i_slint_core::properties::BindingHolder<tuple$<>> * * binding) Line 734	Rust
 	win32_popup_problem.exe!i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>>::for_each<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>,i_slint_core::properties::mark_dependencies_dirty::closure_env$0>(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 self) Line 195	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty(i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>> * dependencies) Line 735	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty::closure$0(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 *, i_slint_core::properties::BindingHolder<tuple$<>> * * binding) Line 734	Rust
 	win32_popup_problem.exe!i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>>::for_each<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>,i_slint_core::properties::mark_dependencies_dirty::closure_env$0>(i_slint_core::properties::mark_dependencies_dirty::closure_env$0 self) Line 195	Rust
 	win32_popup_problem.exe!i_slint_core::properties::mark_dependencies_dirty(i_slint_core::properties::dependency_tracker::DependencyListHead<ptr_const$<i_slint_core::properties::BindingHolder<tuple$<>>>> * dependencies) Line 735	Rust
 	win32_popup_problem.exe!i_slint_core::properties::PropertyHandle::mark_dirty() Line 684	Rust

````

As far as I can tell, when Slint gets frozen as per this bug report, `mark_dirty()` still seems to be called, but `i_slint_core::window::impl$2::notify()` never gets called.  The code between these two points appears to be some sort of property engine, presumably the property engine at the core of the Slint language runtime.  It isn't clear to me how one would go about debugging/diagnosing a problem in this area.

**Comment by ogoffart** [comment]
The mark_dependencies_dirty might not be called if the dependency is already dirty.
It is possible that a call to request_redraw was done before, but then, somehow, the window was not redrawn (and is therefore still dirty), probably because for some reason, the window manager didn't redraw the window despite we asked for it. Something like that.

**Comment by npwoods** [comment]
Yes that was what I was thinking - exercising the demo, you can see that even if the window fails to update, the UI is still active.  Pressing the button will bump the integer, and the new value shows up if the window is resized.

I strongly suspect that in my demo when `TrackPopupMenu` is invoked, a Win32 Message is sent to the `WndProc` of the main window, but Slint is in some state where it does not expect to be called reentrantly.

**Comment by npwoods** [comment]
I discovered that a far simpler way than the `SetWindowPos()` gymnastics to unmute Slint's UI is to invoke the following code:

````
with_winit_window(|x| x.request_redraw());
````

Curiously, the code inside `winitwindowadapter.rs` that is critical to unmute Slint's UI is the following; stepping over it in the debugger keeps Slint muted:
````
renderer.render(self.window())?;
````

Is there a possibility that Slint may be getting into a state where Slint is expecting the window to be redrawn - or more specifically anticipating a `winit` `WindowEvent::RedrawRequested` event - when one is not in fact coming?  I can't speak for other platforms, I am aware that on Windows having `WM_PAINT` (the Win32 message sent to repaint a window) drive behavior beyond simply redrawing the window is discouraged, for reasons (but not limited to) [this](https://devblogs.microsoft.com/oldnewthing/20111219-00/?p=8863) and the kafka-esque nature of the circumstances that can drive the dispatch of `WM_PAINT` messages.

I have noticed that I have noticed that when Slint is mute, `WinitWindowAdapter::pending_redraw` is `true`, though it seems to be insufficient to simply ignore the check in `request_redraw()`

**Comment by npwoods** [comment]
So, a question:  Why do changes that affect the appearance of Slint controls not always result in an invocation of [Window::redraw_requested](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.request_redraw)?

Assuming that other platforms behave similarly to Windows (yes, this is an "if"), the system is smart enough to aggregate/combine the requests to repaint, and potentially reorder the event queue as a consequence.

**Comment by npwoods** [comment]
@ogoffart i see this issue has the "needs info" tag; is there any more information I can provide.

While I'm not convinced I fully understand the issue, I think Slint gets into a state where it is expecting a `winit` `WindowEvent::RedrawRequested` event, but it never comes in practice.

**Comment by ogoffart** [comment]
This seems to be the same issue as https://github.com/slint-ui/slint/issues/8592
This can be reproduced with a basic winit + muda example.
